### ISSUE TITLE
[Bug] Memory leak when using `set_train_data` with autograd enabled

### ISSUE THREAD

**Issue by fteufel** [cornellius-gp/gpytorch#2422]
# 🐛 Bug

Hi, I'm working on a problem that requires me to temporarily add data to a GP's training data in the forward pass. It seems to work when running with `torch.no_grad()`, but memory usage grows with each repeated call when gradients are enabled.

I've removed most of the actual operations I want to do and arrived at the following minimal example that causes a memory leak. In the forward pass, I set the GP's train data to it's own train data - my expectation would be that that costs no memory at all.

## To reproduce

```python

    # this is the forward method of a nn.Module that holds a trained ExactGP as self.model
    def forward(self, X: torch.Tensor) -> torch.Tensor:

        with settings.detach_test_caches(False):
            self.model.eval()

            f_preds = self.model(X) 
            posterior_means = f_preds.mean

            # augment the training data with the test points
            X_train_original = self.model.train_inputs[0]
            Y_train_original = self.model.train_targets
            self.model.set_train_data(X_train_original, Y_train_original, strict=False)

            return torch.sum(posterior_means, dim=1)
```

I call this nn.module repeatedly like
```python
def debug_memory_usage(acquisition_function, candidates, no_grad=True, do_backward=False):
    '''Repeatedly evaluate to check for memory leaks.'''

    if not no_grad:
        for i in range(1000):
            print(f'debug step {i}, memory {torch.cuda.memory_allocated() / 1e9} GB')
            i += 1
            with torch.no_grad():
                X = candidates.requires_grad_(True) # candidates are some torch.rand points

            loss = -dummy_model(X).sum()
            if do_backward:
                grad = torch.autograd.grad(loss, X)[0]

    else:
        with torch.no_grad():
            for i in range(1000):
                print(f'debug step {i}, memory {torch.cuda.memory_allocated() / 1e9} GB')
                i += 1
                X = candidates

                loss = -dummy_model(X).sum()
```

```
Debugging memory usage without gradient computation.
debug step 0, memory 0.000400896 GB
debug step 1, memory 0.008921088 GB
debug step 2, memory 0.008921088 GB
debug step 3, memory 0.008921088 GB
debug step 4, memory 0.008921088 GB
debug step 5, memory 0.008921088 GB
...
debug step 998, memory 0.008921088 GB
debug step 999, memory 0.008921088 GB

Debugging memory usage with gradient enabled - with backward computation.
debug step 0, memory 0.008920576 GB
debug step 1, memory 0.02331136 GB
debug step 2, memory 0.024221184 GB
debug step 3, memory 0.02553088 GB
debug step 4, memory 0.026040832 GB
debug step 5, memory 0.026950656 GB
...
debug step 997, memory 0.929496064 GB
debug step 998, memory 0.930405888 GB
debug step 999, memory 0.931315712 GB
```

## Expected Behavior

As seen in the log without gradients enabled, memory consumption stays constant, as it should. I would expect that to be the same with gradients enabled. Is there an operation i'm missing to make this work? My aim is to be able to add any X to the train data of the GP temporarily in `forward` without it causing a leak.

## System information

**Please complete the following information:**
- GPyTorch 1.10
- PyTorch 2.0.1+cu117
- OS Red Hat Enterprise Linux release 8.8 (Ootpa)
